### PR TITLE
Multi language compilation

### PR DESCRIPTION
**Pull request by AZWN**
This PR adds compilation for projects that use multi-language analysis. It can be activated by setting the `multilangAnalyzer` field on the `LanguageProjectSettings` and `AdapterProjectSettings` of the respective projects.

An example of its usage can be found at [this branch](https://github.com/AZWN/spoofax-pie/tree/generated-multilang-analysis-example)

### CODE REVIEW

**Comment by AZWN** [comment]
Notice that this PR also contains the changes of #5.

File: core/statix.multilang.eclipse/src/main/java/mb/statix/multilang/eclipse/AnalysisContextInitializer.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import mb.spoofax.eclipse.SpoofaxPlugin;
+import mb.statix.multilang.AnalysisContextService;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IExtensionRegistry;
+
+import java.util.stream.Stream;
+
+public class AnalysisContextInitializer {
+    private static final String ANALYSIS_CONTEXT_ID = "mb.metaborg.statix.multilang.analysiscontext";
```
Comment (Gohla) [inline core/statix.multilang.eclipse/src/main/java/mb/statix/multilang/eclipse/AnalysisContextInitializer.java:13]: `mb` == `metaborg`, so `mb.metaborg` is double. Also, we only use `mb` in packages, not in other identifiers, so `mb.metaborg` needs to be removed. There's multiple places where this happens.

File: core/statix.multilang/src/main/java/mb/statix/multilang/utils/MessageUtils.java [hunk 2]
```diff
@@ -110,26 +132,41 @@ private static String cleanupString(String string) {
...
+        if(attachment == null) {
+            return null;
+        }
+        if(attachment instanceof ResourceKeyAttachment) {
+            return ((ResourceKeyAttachment)attachment).resourceKey;
+        }
+
+        if(attachment instanceof ParentAttachment) {
+            IStrategoTerm parentTerm = ((ParentAttachment)attachment).getParent();
+            ISimpleTerm root = ParentAttachment.getRoot(parentTerm);
```
Comment (Gohla) [inline core/statix.multilang/src/main/java/mb/statix/multilang/utils/MessageUtils.java:153]: There seems to be some code duplication with ResourceKeyAttachment (getResourceKey) here.

File: core/spoofax.compiler/src/main/resources/mb/spoofax/compiler/spoofaxcore/adapter_project/Instance.java.mustache [hunk 3]
```diff
@@ -134,4 +167,40 @@ public class {{genInstance.id}} implements LanguageInstance {
...
+                } catch(IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            })
+            .astFunction({{preStatixInjection.name}}.createFunction().mapInput((exec, key) -> {{indexAstInjection.name}}
+                .createSupplier(new {{indexAstInjection.type.qualifiedId}}.Input(key, {{parseInjection.name}}
+                    .createNullableRecoverableAstSupplier(key)))))
+            .postTransform({{postStatixInjection.name}}.createFunction())
+            .languageId(new LanguageId("{{languageId}}"))
+            .addAllTaskDefs(taskDefs)
```
Comment (Gohla) [inline core/spoofax.compiler/src/main/resources/mb/spoofax/compiler/spoofaxcore/adapter_project/Instance.java.mustache:198]: Is it needed to pass the task definitions? Is it possible to use this language's Pie object as a base instead?

File: core/statix.multilang/src/main/java/mb/statix/multilang/utils/MessageUtils.java (hunk 2 above)
Comment (AZWN) [reply core/statix.multilang/src/main/java/mb/statix/multilang/utils/MessageUtils.java:153]: Not exactly. This code also traverses the `ImploderAttachment`. I suspect this is needed because the terms this code is called on are transformed to nabl terms, and then back again.

**Comment by AZWN** [comment]
I think this feature is finished. It grew out to quite a big bunch of code. Tomorrow I will try to  describe the various changes to make reviewing easier.

**Comment by AZWN** [comment]
As promised, I would give an overview of all the changes performed in this PR, to help in reviewing. 

### Utilities

#### Result collector

As discussed, I provided a `java.util.collector` implementation that transforms a `Stream<Result<T, E>>` to a `Result<Set<T>, E>>`. This code lives in the `mb.common.result` package. 

In order to maintain all information from the `Stream`, I introduced an `AggegateException`, which maintains all exceptions _and values_ in the `Stream`.

Moreover, the collector has options to specify a custom base exception (to which all exceptions in the stream are added as suppressed), and a custom collection type (defaulting to `HashSet`).

#### Pie provider

In order to support building multi-language `Pie` instances I created an `PieProvider` indirection, which returns a `Pie` instance based on the project. The default implementation just returns the `Pie` for the language and the implementation used for multi-language analysis tries to read the language configuration, and returns a `Pie` with all the language `Pie`s as ancestor.

In order to prevent accidental `Pie` injections, I qualified the relevant Provider with `@Named("prototype")`.

### Multi-language analysis API

Based on the feedback on the previous implementation, I changed the API quite a bit:

- Firstly, the `AnalysisContext` is removed. All context data is now implicit in the task outputs.
- Secondly, the `AnalysisContextService` now only serves static (default) data. The data it provides can be overridden by the `multilang.yaml` config. The final configuration is provided by the `SmlReadConfigYaml` task definition.
- Thirdly, the api `AnalysisContextService` provides is split into some interfaces, which are injected in the task definitions. This ensures loose coupling between the tasks and the service. I decided not to use a supplier for that, because the `AnalysisContextService` cannot be made serializable.

#### Task definitions

The multilang component provides task definitions to do incremental analysis. I think the definitions are quite self-explanatory. Regarding the task inputs, I only used `LanguageId`s, resource paths (to identify projects), resource keys (for source files) and suppliers based on those keys. Those keys should be small, and prevent task reinstantiation as much as possible.

Two abstract task definitions are provided (`SmlCheckTaskDef` and `SmlIndexAstTaskDef`). These should be implemented by each language. All other tasks can be used as provided.

#### Exception handling

In the tasks, I catch virtually all exceptions, and report them as error messages (using the `MultiLangAnalysisException.toKeyedMessages()`). I tried to make those messages as sensible as possible, by avoiding the generation of messages for exceptions that only wrap other exceptions in their `cause` or `suppressedException` fields.

#### Dagger usage

The `MultiLangComponent` provides the `AnalysisContextService` and the (not-language specific) task definitions with `@MultiLang` qualifier.

The `MultiLangModule` takes a `Supplier<AnalysisContextService>` as parameter. This indirection is needed to prevent circular dependency issues (specifically: To initialize the `AnalysisContextService`, language specific data is needed. That requires the language components to be initialized, but they again depend on the to-be-initialized `MultiLangComponent`).

Moreover, in all task definitions, the `AnalysisContextService` is injected as a `Lazy`, for the same reasons.

### Multi-language analysis in Eclipse

Multi-language analysis in eclipse is provided using the `MultiLangPlugin`. This plugin has the following tasks:
- Ensure languages can reference a `MultiLangComponent`.
- Notify registered languages of changes to `multilang.yaml` config, so that they can update the analysis.

The plugin initializes the `AnalysisContextService` by fetching all language-specific data using an extension point.

Finally, the `MultilangEclipseComponent` provides a shared startup lock, because plugins with shared analysis starting simultaneously give concurrency errors. Eclipse plugins for languages with multilang analysis inject this lock, instead of creating their own.

### Multi-language compilation

Finally, the spoofax3 compiler is extended with a `MultiLangAnalyzerCompiler`. This analyzer ensure that:
- Language-specific versions of the 2 task templates are generated
- The taskdefs are injected properly at all relevant places.
- The `LanguageInstance.createCheckTask()` returns the multilang check taskdef
- The language component depends on the multilang component

Eclipse:
- The multilang extension point is extended with a `LanguageMetadataProvider` for this language.

File: core/statix.multilang.eclipse/src/main/java/mb/statix/multilang/eclipse/ConfigResourceChangeListener.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import org.eclipse.jface.util.SafeRunnable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * {@link IResourceChangeListener} that listens to changes to {@code multilang.yaml} files, and notifies all registered
+ * {@link ConfigChangeListener config change listeners}.
+ */
+public class ConfigResourceChangeListener implements IResourceChangeListener {
```
Comment (Gohla) [inline core/statix.multilang.eclipse/src/main/java/mb/statix/multilang/eclipse/ConfigResourceChangeListener.java:26]: I'm a bit confused as to why this is needed. Didn't we agree that all config file parsing should be done in PIE tasks, so that PIE does the change tracking for us?

File: core/statix.multilang.eclipse/src/main/java/mb/statix/multilang/eclipse/UpdateAnalysisConfigChangeListener.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import mb.spoofax.eclipse.SpoofaxPlugin;
+import mb.spoofax.eclipse.pie.PieRunner;
+import mb.spoofax.eclipse.resource.EclipseResourcePath;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IProgressMonitor;
+
+import java.util.Collections;
+
+public class UpdateAnalysisConfigChangeListener implements ConfigChangeListener {
```
Comment (Gohla) [inline core/statix.multilang.eclipse/src/main/java/mb/statix/multilang/eclipse/UpdateAnalysisConfigChangeListener.java:18]: Same as earlier comment, why is this needed if PIE tasks track the config file?

File: core/statix.multilang/src/main/java/mb/statix/multilang/SharedPieProvider.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public Pie getPie(@Nullable ResourcePath projectDir) {
+        ContextPieManager contextPieManager = this.contextPieManager.get();
+        try(MixedSession session = languagePie.newSession()) {
+            if(projectDir == null) {
+                ContextId defaultContextId = contextPieManager.getDefaultContextId(languageId);
+                Set<LanguageId> languageIds = contextPieManager.getContextLanguages(defaultContextId);
+                return contextPieManager.buildPieForLanguages(languageIds);
+            }
+            Set<LanguageId> languageIds = session.require(buildContextConfiguration
```
Comment (Gohla) [inline core/statix.multilang/src/main/java/mb/statix/multilang/SharedPieProvider.java:49]: Why collect language IDs here? Shouldn't the multilang analysis task of a language require a task that collects the language ids, such that a change in the configuration triggers re-analysis?

File: core/statix.multilang/src/main/java/mb/statix/multilang/SharedPieProvider.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        SmlBuildContextConfiguration buildContextConfiguration
+    ) {
+        this.languagePie = languagePie;
+        this.languageId = languageId;
+        this.contextPieManager = contextPieManager;
+        this.buildContextConfiguration = buildContextConfiguration;
+    }
+
+    @Override
+    public Pie getPie(@Nullable ResourcePath projectDir) {
```
Comment (Gohla) [inline core/statix.multilang/src/main/java/mb/statix/multilang/SharedPieProvider.java:41]: I really dislike that a project path is used to create a PIE instance. Getting a PIE instance should be decoupled from project paths, or any kinds of paths. Only tasks of a language that require a project path should have access to it.

**Review by Gohla — CHANGES_REQUESTED** [review]

**Comment by AZWN** [comment]
I updated the branch with the interface dependency feature. The updated example can be found [here](https://github.com/AZWN/spoofax-pie/tree/generated-multilang-analysis-example/example/multilang).

**Review by Gohla — APPROVED** [review]
